skbio/metadata/_interval.py

```diff
@@ -0,0 +1,465 @@
+        self._intervals = IntervalTree()
+        self._is_stale_tree = False
+
+    def _reverse(self, length):
+        """ Reverse complements IntervalMetadata object.
+
+        This operation reverses all of the interval coordinates.
+        For instance, this can be used to compare coordinates
+        in the forward strand to coordinates in the reversal strand.
+
+        Parameters
+        ----------
+        length : int
+            Largest end coordinate to perform reverse complement.
+            This typically corresponds to the length of sequence.
+        """
+        for f in self._metadata:
+            # staled doesn't need to be called, since the setter for
+            # Interval will take care of this
+            invs = list(map(lambda x: (length-x[1], length-x[0]),
+                            f.intervals))
+            f.intervals = invs
+
+    def add(self, intervals, boundaries=None, metadata=None):
+        """ Adds a feature to the metadata object.
+
+        Parameters
+        ----------
+        intervals : iterable of tuple of ints
+            A list of intervals associated with the Interval object.
+        boundaries : iterable of tuple of bool
+            A list of boundaries associated with the Interval object.
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            Interval object.
+
+        Examples
+        --------
+        >>> from skbio.metadata import IntervalMetadata
+        >>> interval_metadata = IntervalMetadata()
+        >>> interval_metadata.add(intervals=[(0, 2)],
+        ...                       boundaries=None, metadata={'name': 'sagA'})
+        >>> interval_metadata.add(intervals=[(40, 70)],
+        ...                       boundaries=None, metadata={'name': 'sagB'})
+        >>> list(interval_metadata.query(intervals=[(1, 2)]))
+        [<Interval: start=0, end=2, 1 contiguous interval, 1 metadata keys, \
+dropped=False>]
+
+        """
+        # Add an interval to the tree. Note that the add functionality is
+        # built within the Interval constructor.
+        Interval(interval_metadata=self,
+                 intervals=intervals,
+                 boundaries=boundaries,
+                 metadata=metadata)
+
+    def _rebuild_tree(self, intervals):
+        """ Rebuilds the IntervalTree when the tree is stale."""
+        self._intervals = IntervalTree()
+        for f in intervals:
+            for start, end in f.intervals:
+                self._intervals.add(start, end, f)
+
+    def _query_interval(self, interval):
+        """ Fetches Interval objects based on query interval"""
+        _assert_valid_interval(interval)
+        start, end = interval
+        invs = self._intervals.find(start, end)
+        return invs
+
+    def _query_attribute(self, intervals, metadata):
+        """ Fetches Interval objects based on query attributes"""
+        if metadata is None:
+            return []
```

Comment:
This should just be ``return`` - the value will never get used for anything. For example:

```python
In [1]: def f():
   ...:     return "hello"
   ...:     yield 42


In [2]: list(f())
Out[2]: []
```